usso/openid.go

```diff
@@ -0,0 +1,103 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017-2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package usso
+
+import (
+	"html/template"
+	"net/http"
+	"strings"
+
+	"github.com/CanonicalLtd/serial-vault/datastore"
+	"github.com/juju/usso"
+	"github.com/juju/usso/openid"
+)
+
+var (
+	// Teams are hardcoded and not currently used.
+	// The team config is here for reference only, but could be used in the future
+	teams    = "ce-web-logs,canonical"
+	required = "email,fullname,nickname"
+	optional = ""
+)
+
+var client = openid.NewClient(usso.ProductionUbuntuSSOServer, &datastore.OpenidNonceStore, nil)
+
+// verify is used to perform the OpenID verification of the login
+// response. This is declared as a variable so it can be overridden for
+// testing.
+var verify = client.Verify
```

Comment:
I'm not really sure that I understand this in this particular case. If you look at the implementation of the openid library, they are using the same approach: https://github.com/juju/usso/blob/master/openid/openid.go#L182

The reason to only mock Verify is that it does a callback to the USSO service to check that the details are from the real server. So this is just mocking that callback.